lib/src/constants.dart

```diff
@@ -0,0 +1,171 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/constant/value.dart';
+
+/// Always returns `null` (used as a default for `defaultTo` methods).
+Null _alwaysNull() => null;
+
+/// Returns whether or not [object] is or represents a `null` value.
+bool _isNull(DartObject object) => object?.isNull != false;
+
+/// Similar to [DartObject.getField], but traverses super classes.
+DartObject _getFieldRecursive(DartObject object, String field) {
+  if (_isNull(object)) {
+    return null;
```

Comment:
If you don't mind I'm addressing that in the next PR - the code to check all fields is a little meaty.

Filed https://github.com/dart-lang/source_gen/issues/173.